File: implementations/cpp/dijkstra.cpp
#include <algorithm>
#include <cmath>
#include <functional>
#include <iostream>
#include <limits>
#include <optional>
#include <queue>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

// Trusted, repository-controlled Dijkstra reference implementation.
using NodeId = std::string;

struct Edge {
  NodeId to;
  double weight;
};

using Graph = std::unordered_map<NodeId, std::vector<Edge>>;

struct Result {
  bool found;
  std::optional<double> cost;
  std::vector<NodeId> path;
};

void validate_graph(const Graph& graph, const NodeId& start, const NodeId& goal) {
  if (!graph.contains(start) || !graph.contains(goal)) {
    throw std::invalid_argument("start and goal must be present in the graph");
  }
  for (const auto& [node, edges] : graph) {
    (void)node;
    for (const Edge& edge : edges) {
      if (!graph.contains(edge.to)) {
        throw std::invalid_argument("unknown node: " + edge.to);
      }
      if (!std::isfinite(edge.weight) || edge.weight < 0) {
        throw std::invalid_argument("Dijkstra requires finite, nonnegative weights");
      }
    }
  }
}

Result dijkstra(const Graph& graph, const NodeId& start, const NodeId& goal) {
  validate_graph(graph, start, goal);

  std::unordered_map<NodeId, double> distances;
  std::unordered_map<NodeId, NodeId> parents;
  for (const auto& [node, edges] : graph) {
    (void)edges;
    distances[node] = std::numeric_limits<double>::infinity();
  }
  distances[start] = 0.0;

  using QueueEntry = std::pair<double, NodeId>;
  std::priority_queue<QueueEntry, std::vector<QueueEntry>,
                      std::greater<QueueEntry>>
      frontier;
  frontier.emplace(0.0, start);

  while (!frontier.empty()) {
    const auto [current_distance, current] = frontier.top();
    frontier.pop();
    if (current_distance != distances.at(current)) continue;
    if (current == goal) break;

    for (const Edge& edge : graph.at(current)) {
      const double candidate = current_distance + edge.weight;
      if (candidate >= distances.at(edge.to)) continue;

      distances[edge.to] = candidate;
      parents[edge.to] = current;
      frontier.emplace(candidate, edge.to);
    }
  }

  if (std::isinf(distances.at(goal))) return {false, std::nullopt, {}};

  std::vector<NodeId> path;
  for (NodeId node = goal;; node = parents.at(node)) {
    path.push_back(node);
    if (node == start) break;
  }
  std::reverse(path.begin(), path.end());
  return {true, distances.at(goal), path};
}

int main() {
  const Graph graph = {
      {"A", {{"B", 2}, {"C", 7}}},
      {"B", {{"C", 1}, {"D", 5}}},
      {"C", {{"D", 1}}},
      {"D", {}},
  };
  const Result result = dijkstra(graph, "A", "D");
  const std::vector<NodeId> expected = {"A", "B", "C", "D"};
  if (!result.found || result.cost != 4.0 || result.path != expected) return 1;
  std::cout << "cpp dijkstra: ok\n";
}

File: implementations/python/dijkstra.py
"""Trusted, repository-controlled Dijkstra reference implementation."""

from __future__ import annotations

from dataclasses import dataclass
import heapq
import math
from typing import Mapping, Sequence


NodeId = str
Graph = Mapping[NodeId, Sequence[tuple[NodeId, float]]]


@dataclass(frozen=True)
class DijkstraResult:
    found: bool
    cost: float | None
    path: list[NodeId]


def _validate_graph(graph: Graph, start: NodeId, goal: NodeId) -> None:
    if start not in graph or goal not in graph:
        raise ValueError("start and goal must be present in the graph")
    for edges in graph.values():
        for neighbor, weight in edges:
            if neighbor not in graph:
                raise ValueError(f"unknown node: {neighbor}")
            if not math.isfinite(weight) or weight < 0:
                raise ValueError("Dijkstra requires finite, nonnegative weights")


def dijkstra(graph: Graph, start: NodeId, goal: NodeId) -> DijkstraResult:
    """Return one deterministic shortest path from start to goal."""

    _validate_graph(graph, start, goal)
    distances = {node: math.inf for node in graph}
    parents: dict[NodeId, NodeId] = {}
    distances[start] = 0.0

    frontier: list[tuple[float, NodeId]] = [(0.0, start)]
    while frontier:
        current_distance, current = heapq.heappop(frontier)
        if current_distance != distances[current]:
            continue
        if current == goal:
            break

        for neighbor, weight in graph[current]:
            candidate = current_distance + weight
            if candidate >= distances[neighbor]:
                continue

            distances[neighbor] = candidate
            parents[neighbor] = current
            heapq.heappush(frontier, (candidate, neighbor))

    if math.isinf(distances[goal]):
        return DijkstraResult(found=False, cost=None, path=[])

    path = [goal]
    while path[-1] != start:
        path.append(parents[path[-1]])
    path.reverse()
    return DijkstraResult(found=True, cost=distances[goal], path=path)


def _smoke_test() -> None:
    graph: Graph = {
        "A": [("B", 2), ("C", 7)],
        "B": [("C", 1), ("D", 5)],
        "C": [("D", 1)],
        "D": [],
    }
    result = dijkstra(graph, "A", "D")
    assert result == DijkstraResult(True, 4.0, ["A", "B", "C", "D"])


if __name__ == "__main__":
    _smoke_test()
    print("python dijkstra: ok")

